### PR TITLE
add getAllMessages subscription

### PR DESCRIPTION
**Pull request by seekdavidlee**
The following are events that can be consumed from the ChatList component.

```ts
    | 'chatMessageReceived'
    | 'chatMessageDeleted'
    | 'chatMessageEdited'
    | 'chatRenamed'
    | 'memberAdded'
    | 'memberRemoved'
```

The `StatefulGraphChatListClient.ts` and `GraphChatListClient` are copied from `StatefulGraphChatClient.ts` and `GraphChatClient` respectively, but refactored to support user versus chat context. `ChatMessage` is the raw type returned back from the subscription and this can be consumed by the ChatList component.

You may also notice `StatefulGraphChatListClient` exposes 2 types of subscriptions. `onStateChange` is for exposing changes to its own state. This is the original use-case for `StatefulGraphChatClient` and I kept it because we needed to get the user Id. `onChatListEvent` is used to consume message, member, chat renamed events.

### CODE REVIEW

**Comment by plasne** [comment]
I would suggest we fix the double-subscription issue before we merge this:

image

File: packages/mgt-chat/src/statefulClient/StatefulGraphChatListClient.ts [hunk 1]
```diff
@@ -0,0 +1,772 @@
...
+  public onStateChange(handler: (state: GraphChatListClient) => void): void {
+    if (!this._stateSubscribers.includes(handler)) {
+      this._stateSubscribers.push(handler);
+    }
+  }
+
+  /**
+   * Unregister a callback from receiving state updates
+   *
+   * @param {(state: GraphChatClient) => void} handler
```
Comment (TechnicallyWilliams) [inline packages/mgt-chat/src/statefulClient/StatefulGraphChatListClient.ts]: Here too

```suggestion
   * @param {(state: GraphChatListClient) => void} handler
```

File: packages/mgt-chat/src/statefulClient/StatefulGraphChatListClient.ts [hunk 2]
```diff
@@ -0,0 +1,772 @@
...
+    const index = this._messageSubscribers.indexOf(handler);
+    if (index !== -1) {
+      this._messageSubscribers = this._messageSubscribers.splice(index, 1);
+    }
+  }
+
+  /**
+   * Register a callback to receive state updates
+   *
+   * @param {(state: GraphChatClient) => void} handler
```
Comment (TechnicallyWilliams) [inline packages/mgt-chat/src/statefulClient/StatefulGraphChatListClient.ts]:
```suggestion
   * @param {(state: GraphChatListClient) => void} handler
```

Comment (seekdavidlee) [reply packages/mgt-chat/src/statefulClient/StatefulGraphChatListClient.ts]: Thank you Dex. I have updated.

File: packages/mgt-chat/src/components/ChatList/ChatList.tsx [hunk 3]
```diff
@@ -42,42 +43,60 @@ export const ChatList = (
       buttonItems?: ChatListButtonItem[];
     }
 ) => {
+  const { value } = props.dataContext as { value: GraphChat[] };
+  const chats: GraphChat[] = value;
```
Comment (kavithasreenath) [inline packages/mgt-chat/src/components/ChatList/ChatList.tsx:56]: isn't this assignment redundant? Can we not use value (renamed as chats in the destructuring assignment) directly in the useState later on?

**Review by plasne — APPROVED** [review]

**Review by TechnicallyWilliams — APPROVED** [review]

**Pull request merged**